bot/src/main.rs

```diff
@@ -2442,29 +2442,41 @@ fn generate_csv(epoch: Epoch, config: &Config) -> BoxResult<()> {
 }
 
 /// Given a validator's current commission and history of commission changes, returns the validator's commission at the end of `epoch`
+/// Only works if the commission change history includes all changes for `epoch` and `epoch + 1`.
 fn calculate_commission_at_end_of_epoch(
     epoch: u64,
     current_commission: u8,
     commission_change_history: Option<&Vec<CommissionChangeIndexHistoryEntry>>,
 ) -> u8 {
     match commission_change_history {
         Some(records) => {
-            let mut rs: Vec<&CommissionChangeIndexHistoryEntry> = records
-                .iter()
-                .filter(|r| r.commission_before.is_some() && r.epoch == epoch + 1)
-                .collect();
-
-            if rs.is_empty() {
-                current_commission
+            // First check if there is a commission change record in `epoch`. The last one will
+            // give us the commision at the end of the epoch.
+            let mut rs: Vec<&CommissionChangeIndexHistoryEntry> =
+                records.iter().filter(|r| r.epoch == epoch).collect();
+
+            if !rs.is_empty() {
+                rs.sort_by(|a, b| a.epoch_completion.partial_cmp(&b.epoch_completion).unwrap());
+                info!("{:?}", rs);
+                rs.last().unwrap().commission_after as u8
             } else {
-                rs.sort_by(|a, b| {
-                    a.epoch
-                        .cmp(&b.epoch)
-                        .then(a.epoch_completion.partial_cmp(&b.epoch_completion).unwrap())
-                });
-                rs.first().unwrap().commission_before.unwrap() as u8
+                // If we didn't find a commission change in `epoch`, check for commission changes in
+                // `epoch + 1`. The first one will give us the commission at the end of `epoch`.
+                let mut rs: Vec<&CommissionChangeIndexHistoryEntry> = records
+                    .iter()
+                    .filter(|r| r.commission_before.is_some() && r.epoch == epoch + 1)
+                    .collect();
+                if rs.is_empty() {
+                    // no commission changes in epoch `epoch + 1`; commission is the current
+                    // commission.
+                    current_commission
+                } else {
+                    rs.sort_by(|a, b| a.epoch_completion.partial_cmp(&b.epoch_completion).unwrap());
+                    rs.first().unwrap().commission_before.unwrap() as u8
```

Comment:
Is this a safety check in case the bot doesn't run in the future? I'm trying to understand better why we'd want the commission coming from the middle of the epoch in which the bot is running.

Reply:
Here's an example:

`epoch` = 123 and the current epoch is 124. 
Current commission is 10%
Epoch change history:
- from 10% to 100% in epoch 120 (`epoch - 3`) (alternately, there could be no record because the last commission change was before we have records of commission changes)
- from 100% to 10% in epoch 124 (`epoch + 1`)

In this case, to get the commission at the end of 123 we need to look at the commission change that happened in 124.

Thinking this through, this does make me realize that we don't need to look at just 124 in this case... any commission change in epochs > 123 are relevant.